src/SimpleMeshing/src/mesh.jl

```diff
@@ -244,5 +245,61 @@ end
     return head, tail
 end
 
+using MPI
+
+struct BufferedArray{T, N, A<:AbstractArray{T,N}} <: AbstractArray{T,N}
+    arr::A
+    recv_buffers::Tuple
+    send_buffers::Tuple
+end
+
+Base.size(a::BufferedArray) = size(a.arr)
+Base.getindex(a::BufferedArray, idx...) = a.arr[idx...]
+Base.setindex!(a::BufferedArray, val, idx...) = a.arr[idx...] = val
+Base.axes(a::BufferedArray) = axes(a.arr) # now includes offsets
+Base.IndexStyle(a::BufferedArray) = IndexStyle(a.arr)
+
+function flush_recvbufs!(a::BufferedArray, mesh::GhostCartesianMesh)
+    for (bidx, buf) in zip(ghostboundaries(mesh), a.recv_buffers)
+        a[bidx] .= buf
+    end
+end
+
+function fill_sendbufs!(a::BufferedArray, mesh::GhostCartesianMesh)
+    for (bidx, buf) in zip(ghostboundaries(mesh), a.recv_buffers)
+        buf .= a[bidx]
+    end
+end
+
+function mpistorage(::Type{T}, mesh::GhostCartesianMesh{N, CPU}) where {T, N}
+    recv_bufs = map(idxs -> Array{T}(undef, size(idxs)), ghostboundaries(mesh))
+    send_bufs = map(idxs -> Array{T}(undef, size(idxs)), ghostboundaries(mesh))
+    BufferedArray(storage(T, mesh), recv_bufs, send_bufs)
+end
+
+const mpicomm = MPI.COMM_WORLD
+
+function sync!(a::BufferedArray, mesh, nbrranks)
+    # TODO: reorder this to first receive and then send
+
+    fill_sendbufs!(a, mesh)
+
+    sendreqs = fill(MPI.REQUEST_NULL, length(nbrranks))
```

Comment:
`sendreqs` and `recvreqs` need to be part of BufferedArray, otherwise `sync; sync;` will do the wrong thing.